### PR TITLE
init creates a symlink to actual repo directory.

### PR DESCRIPTION
**Pull request by anorth**
The init command creates a symbolic link to a repo directory.
Updates lots of tests, especially the migration ones which necessarily inspected the filesystem.

### CODE REVIEW

File: repo/fsrepo.go [hunk 1]
```diff
@@ -431,12 +489,15 @@ func isEmptyDir(path string) (bool, error) {
 	return len(infos) == 0, nil
 }
 
-func fileExists(file string) bool {
+func fileExists(file string) (bool, error) {
```
Comment (shannonwells) [inline repo/fsrepo.go]: Is this functionality not anywhere else in the code base?  (for that matter why isn't this entire thing part of golang, but that's another issue)

**Comment by anorth** [comment]
@ZenGround0 @shannonwells this is based on https://github.com/filecoin-project/go-filecoin/pull/2701 but otherwise ready for review. Sorry it got a bit long with the test updates.

File: commands/init.go [hunk 2]
```diff
@@ -11,7 +11,7 @@ import (
 
 	"github.com/ipfs/go-car"
 	hamt "github.com/ipfs/go-hamt-ipld"
-	"github.com/ipfs/go-ipfs-blockstore"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
```
Comment (shannonwells) [inline commands/init.go]: remove unneeded import name

File: tools/migration/internal/repo_fs_helpers.go [hunk 3]
```diff
@@ -17,73 +17,68 @@ import (
...
+	now := time.Now()
 	var newpath string
-	for i := 1; i < 1000; i++ {
-
-		now := time.Now()
-		nowStr := now.Format("20060102-150405")
-
-		newpath = strings.Join([]string{oldPath, nowStr, fmt.Sprintf("%03d", i)}, "-")
+	for i := uint(0); i < 1000; i++ {
+		newpath = repo.MakeRepoDirName(linkPath, now, version, i)
```
Comment (shannonwells) [inline tools/migration/internal/repo_fs_helpers.go]: 🎉 🎈 just checking

File: repo/fsrepo.go [hunk 4]
```diff
@@ -123,22 +173,37 @@ func OpenFSRepo(repoPath string) (*FSRepo, error) {
...
+		name = name + fmt.Sprintf("-%d", uniqueifier)
+	}
+	return name
+}
+
 func (r *FSRepo) loadFromDisk() error {
-	localVersion, err := r.loadVersion()
+	localVersion, err := r.readVersion()
 	if err != nil {
 		return errors.Wrap(err, "failed to load version")
```
Comment (shannonwells) [inline repo/fsrepo.go]: if it's being changed to readVersion should it say "failed to read version"?

File: repo/fsrepo_test.go [hunk 5]
```diff
@@ -82,20 +83,26 @@ const (
...
-		defer func() {
-			require.NoError(t, os.RemoveAll(dir))
-		}()
+		defer RequireRemoveAll(t, dir)
 
-		_, err = initAndOpenRepo(dir, cfg)
+		_, err = initAndOpenRepoDirect(dir, 42, cfg)
 		assert.NoError(t, err)
-		checkNewRepoFiles(t, dir)
+		checkNewRepoFiles(t, dir, 42)
```
Comment (shannonwells) [inline repo/fsrepo_test.go]: "It says, 'what do you get if you multiply six by nine?'"
"Six by nine. Forty two."
"That's it. That's all there is."
"I always thought something was fundamentally wrong with the universe."

**Comment by anorth** [comment]
Ping @ZenGround0 @shannonwells

File: tools/migration/internal/runner_buildonly_test.go [hunk 6]
```diff
@@ -34,18 +35,18 @@ func TestMigrationRunner_RunBuildonly(t *testing.T) {
...
+		runner, err := NewMigrationRunner(logger, "buildonly", repoSymLink, "")
 		require.NoError(t, err)
 		runner.MigrationsProvider = testProviderPasses
 
 		runResult := runner.Run()
 		assert.NoError(t, runResult.Err)
 
-		AssertBumpedVersion(t, runResult.NewRepoPath, repoDir, 0)
+		AssertBumpedVersion(t, runResult.NewRepoPath, repoSymLink, 0)
 		assert.Equal(t, uint(1), runResult.NewVersion)
```
Comment (ZenGround0) [inline tools/migration/internal/runner_buildonly_test.go]: I didn't understand this the first time around either but didn't follow up enough: why are `"writes the new version to the new repo, does not install"` and `"clones repo"` not combined into a single test?  We run "build-only" in both cases.  Both runs should create the new dir, bump the version, and not link the new dir right?

File: tools/migration/internal/test_helpers.go [hunk 7]
```diff
@@ -43,32 +24,41 @@ func AssertLogged(t *testing.T, logFile *os.File, subStr string) {
...
+	err := repo.InitFSRepo(repoLink, version, config.NewDefaultConfig())
+	require.NoError(t, err)
+	return
+}
 
-	require.NoError(t, repo.WriteVersion(repoDir, repoVersion))
-	return repoDir, symLink
+// CaptureOutput redirects log content into a buffer and calls the function provided,
+// and returns buffer content.
+func CaptureOutput(f func()) string {
```
Comment (ZenGround0) [inline tools/migration/internal/test_helpers.go]: Is this left over from a rebase?  I believe this is deadcode now.

File: repo/testing.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+// RequireMakeTempDir ensures that a temporary directory is created
+func RequireMakeTempDir(t *testing.T, dirname string) string {
+	newdir, err := ioutil.TempDir("", dirname)
+	require.NoError(t, err)
+	return newdir
+}
+
+// RequireRemoveAll ensures that the error condition is checked when we clean up
+// after creating a temporary directory.
+func RequireRemoveAll(t *testing.T, path string) {
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply repo/testing.go]: oh duh this is where they moved from

File: repo/fsrepo_test.go [hunk 9]
```diff
@@ -148,71 +151,67 @@ func TestFSRepoOpen(t *testing.T) {
 	tf.UnitTest(t)
 
 	t.Run("[fail] repo version newer than binary", func(t *testing.T) {
-		dir, err := ioutil.TempDir("", "")
+		container, err := ioutil.TempDir("", "")
 		assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (anorth) [reply repo/fsrepo_test.go]: Fixed

File: tools/migration/internal/repo_fs_helpers.go [hunk 10]
```diff
@@ -17,73 +17,68 @@ import (
 //
 // This does touch sector data.
```
[the remark this answers is not in the document]
Comment (anorth) [reply tools/migration/internal/repo_fs_helpers.go]: Probably a mistake when it was written. Fixed.

File: tools/migration/internal/runner_buildonly_test.go (hunk 6 above)
Comment (anorth) [reply tools/migration/internal/runner_buildonly_test.go]: Good catch, done.

**Pull request merged**